new blog post on storage options for corso (#2362)
alcionai/corso: website/blog/2023-2-4-where-to-store-corso.md

---
slug: where-to-store-corso
title: "Where to store your Corso Repository"
description: "Storage Options for Corso"
authors: nica
tags: [corso, microsoft 365, backups, S3]
date: 2023-2-4
image: ./images/boxes_web.jpeg
---

![image of a large number of packing boxes](./images/boxes_web.jpeg)

We all know that Corso is a free and open-source tool for creating backups of your Microsoft 365 data. But where does
that data go?

Corso creates a repository to store your backups, and the default in our documentation is to send that data to AWS S3.
It is possible however to back up to any object storage system that has an S3-compatible API. Let’s talk about some options.

> 

Reviewer: We need a truncate somewhere

Author: added

amount of storage used as well as the amount of network traffic when writing to object storage. Corso also combines
different emails, attachments, etc. into larger objects to make it more cost-effective by reducing the number of API
calls and increasing network throughput as well as making Corso data eligible and cost-effective for some of the other
storage classes described below.

Stepping away from the default S3 offering (S3 Standard), S3 offers a number of different Glacier (cheap and deep)
storage classes that can help to further reduce the cost for backup and archival workloads. Within the storage classes,
Corso today supports Glacier Instant Retrieval but, because of user responsiveness and metadata requirements, not the
other Glacier variants.

Glacier Instant Retrieval should provide the best price performance for a  backup workload as backup data blobs are
typically written once, with occasional re-compacting, and read infrequently in the case of restore. One should note
that recommendations such as these are always workload dependent and should be verified for your use case. For example,

> we would not recommend Glacier IR if you are constantly testing large restores or have heavy churn in your backups and

Reviewer: ```suggestion
we would not recommend Glacier Instant Retrieval if you are constantly testing large restores or have heavy churn in your backups and
```

Author: replaced

---
slug: where-to-store-corso
title: "Where to store your Corso Repository"
description: "Storage Options for Corso"
authors: nica
tags: [corso, microsoft 365, backups, S3]
date: 2023-2-4
image: ./images/boxes_web.jpeg
---

![image of a large number of packing boxes](./images/boxes_web.jpeg)

We all know that Corso is a free and open-source tool for creating backups of your Microsoft 365 data. But where does
that data go?

Corso creates a repository to store your backups, and the default in our documentation is to send that data to AWS S3.
It is possible however to back up to any object storage system that has an S3-compatible API. Let’s talk about some options.

## **S3-Compatible Object Storage**

A number of other cloud providers aren’t the 500-pound gorilla of AWS but still offer an S3-compatible API.
Some of them include:

- Google Cloud: One of the largest cloud providers in the world, Google offers

> 

Reviewer: ```suggestion
```

---
slug: where-to-store-corso
title: "Where to store your Corso Repository"
description: "Storage Options for Corso"
authors: nica
tags: [corso, microsoft 365, backups, S3]
date: 2023-2-4
image: ./images/boxes_web.jpeg
---

![image of a large number of packing boxes](./images/boxes_web.jpeg)

We all know that Corso is a free and open-source tool for creating backups of your Microsoft 365 data. But where does
that data go?

Corso creates a repository to store your backups, and the default in our documentation is to send that data to AWS S3.

> It is possible however to back up to any object storage system that has an S3-compatible API. Let’s talk about some options.

Reviewer: ```suggestion
It's possible however to back up to any object storage system that has an S3-compatible API. Let’s talk about some options.
```

Digging into how Corso works, it will often delete and rewrite the data and
metadata indices within a repository.
That’s how it can offer you great detail into your backups,
like in this video where I look at [individual backed-up emails](https://youtu.be/mweAUDhUE7I) from my exchange server.
As a result of these frequent reads and writes, and the way Corso garbage
collects deleted backups, AWS S3 Glacier Deep Archive or Flexible Retrieval,
with several minutes or longer request latency, won’t work with Corso,
full stop.

However, Glacier Instant Retrieval (IR) is a good candidate to store the objects
that represent the backup data but not the various indexes needed to maintain the
repo.
Glacier Instant Retrieval should provide the best price performance for a backup workload as backup data blobs are
typically written once, with occasional re-compacting, and read infrequently in the case of restore.
One should note that recommendations such as these are always workload dependant and should be verified for your use
case.
For example, we would not recommend Glacier IR if you are constantly testing large restores or have heavy churn in your
backups and limited retention.

> For most typical backup workloads (write mostly, read rarely), Glacier Instant Retrieval should work just fine and
> deliver the best price-performance ratio.

Reviewer: ```suggestion
For most typical workloads, Glacier Instant Retrieval should be a great fit for Corso given how it stores data within its repository.
```

date: 2023-2-4
image: ./images/boxes_web.jpeg
---

![Filled coffee cup by Kenny Louie from Vancouver, Canada -Coffee break, CC BY 2.0, https://commons.wikimedia.org/w/index.php?curid=24336256](./images/boxes_web.jpeg)

We all know that Corso is a free and open-source tool for creating backups of your Microsoft 365 data.
But where does that data go?

Corso creates a repository to store your backups, and the default in our
documentation is to send that data to AWS S3.
It is possible however to back up to any object storage system that has an S3-compatible
API.
Let’s talk about some options.

<!-- truncate -->

## **S3-Compatible Object Storage**

A number of other cloud providers aren’t the 500-pound gorilla of AWS but still
offer an S3-compatible API.
Some of them include:

- Google Cloud: One of the largest cloud providers in the world, Google offers
[an S3-compatible API](https://cloud.google.com/storage/docs/interoperability) on
 top of its Google Cloud Storage (GCS) offering.
- Backblaze: Known for its deep analysis of hard drive failure statistics,
Backblaze offers an S3-compatible API for its B2 Cloud Storage product.
They also make the bold claim of costing

> [significantly less than AWS S3]
> (https://www.backblaze.com/b2/cloud-storage-pricing.html)

Reviewer: ```suggestion
[significantly less than AWS S3](https://www.backblaze.com/b2/cloud-storage-pricing.html)
```

That’s how it can offer you great detail into your backups,
like in this video where I look at [individual backed-up emails](https://youtu.be/mweAUDhUE7I) from my exchange server.
As a result of these frequent reads and writes, and the way Corso garbage
collects deleted backups, AWS S3 Glacier Deep Archive or Flexible Retrieval,
with several minutes or longer request latency, won’t work with Corso,
full stop.

However, Glacier Instant Retrieval (IR) is a good candidate to store the objects
that represent the backup data but not the various indexes needed to maintain the
repo.
Glacier Instant Retrieval should provide the best price performance for a  backup workload as backup data blobs are
typically written once, with occasional re-compacting, and read infrequently in the case of restore.
One should note that recommendations such as these are always workload dependant and should be verified for your use
case.
For example, we would not recommend Glacier IR if you are constantly testing large restores or have heavy churn in your
backups and limited retention.

For most typical backup workloads (write mostly, read rarely), Glacier Instant Retrieval should work just fine and

> deliver the best price-performance ratio.

Reviewer: Let me take a stab here.

> deliver the best price-performance ratio.

Author: resolved by significant edits in final version

Unlike a naive backup solution and as mentioned above, Corso efficiently packs smaller backup items (e.g., multiple
emails) into a larger object.
This makes Corso’s backup data eligible for S3 Intelligent Tiering where infrequently accessed objects can be moved into
Glacier IR.
However, if you have a typical backup workload, Corso also supports putting objects containing backup data directly into
the Glacier IR storage class without waiting for a “cooling off”

> time spent in the more expensive S3 Standard storage class.

Reviewer 2: This section on Intelligent Tiering seems weirdly placed and the transitions from the Glacier IR one is too abrupt. 

I would consider having too distinct sections with headings with recommendations to consider in order of priority
* Glacier IR
* S3 Intellignet tiering

It would be much better if there is content along the lines pros and cons for each or at least examples of what workloads are good for both. 

For Glacier IR you sort of have that - typical backup workload.

When should I consider S3 Intelligent Tiering? - maybe when I not sure about the relative frequency of restores and  my retention policy. 

As is the tiering section is mentioned but it is unclear why I would care because then it says use Glacier IR anyway. 

If you go with the subsection approach, you can also include when to use S3 Standard - it is already in the text - performing testing with large repos where you are churning data rapidly and performing frequent test restores, etc

This is probably more feedback than you asked for so you can chose how to use it

Reviewer: Georgi, I agree this is a lot that gets backed in here and the transitions are awkward. I guess it evolved this way, with contributions from me, from wanting to position this as Glacier in the context of Corso. Your suggestions would improve the article but it is a significant rework (and maybe even its own article?). I will defer to @serverless-mom about what she wants to do.

Author: I think we cut everything about intelligent tiering and perhaps write a follow up piece on 'what S3 storage class is right for backups,' I can get a new version drafted and sent over. Since it's really not intended for prose reviews, Github is not the best place for large rewrites, I'll take this back to Notion and work from there

Author: resolved by rewrites. I think a 'choosing the best AWS storage class for your backups' is a worthwhile blog post topic, will draft
